### ISSUE TITLE
Add a Complex32 datatype

### ISSUE THREAD

**Issue by barrettp** [casangi/xradio#304]
Many radio telescope A2D samplers are 8 bit, e.g., those used by the VLA. This means that the visibility data can be stored in half-float (16-bit) format without loss of precision. Use of half-floats will reduce the storage of such data by about half. It is recommended that half-floats be included as an xarray datatype.

**Comment by kettenis** [comment]
https://github.com/numpy/numpy/issues/14753

**Comment by FedeMPouzols** [comment]
Yes, the fact that numpy does not support a `numpy.complex32` type effectively blocks this. If I remember correctly we initially listed complex32 but then had to remove it because of lack of support in numpy.
As stated in one comment, there are no concrete plans for `numpy.complex32` , and it doesn't appear on the numpy roadmap: https://numpy.org/neps/roadmap.html
I'd also note one would need to double-check performance issues with complex32 / 16-bit floats, depending on CPU or GPU hardware (https://stackoverflow.com/questions/15340781/python-numpy-data-types-performance, https://stackoverflow.com/questions/38975770/python-numpy-float16-datatype-operations-and-float8).

The MSv4 Review Committee gave a comment related to this issue:
```
There were questions about handling sparse data (322.1) and the possibility of 32-bit complex numbers
(304). The response here suggested this will be dealt with using compression during storage, both using
built-in Zarr NaN-filling and data compression and in future with a Dysco compression option. There
should not be any performance or storage space impacts, but this may need to be demonstrated.
```

The MSv4 Review Committee considered this issue/RID closed without further action.

Please re-open or open a related issue if there is anything else to discuss or fix.

**Comment by barrettp** [comment]
You do realize that there are other languages than Python, namely Julia, that support Complex32. Why should other implementations suffer because numpy cannot support that data type? This is not a good approach to software development. XRadio should be language agnostic, not Python-centric.

If you want to test the performance of Complex32, then perform a test using Julia. Many GPUs now support Float16.

**Comment by kettenis** [comment]
A community standard should not require features that are not widely available: it should be possible to read an MSv4 using all widely used language in the community. I posted the link to the numpy issue as it points out several issues with half-precision that are not unique to Python (but it does mean this is currently a showstopper for using Complex32 in the xradio reference implementation).

Not all hardware implements support for half-precision floating-point. Most x86 CPUs only support it as a storage format. This means that to do any calculations, conversion to single-precision floating-point is required. This is why in many languages half-precision is not fully supported. I realize that documentation is not necessarily up to date, but even the current Julia documentation says:
> Half-precision floating-point numbers are also supported ([Float16](https://docs.julialang.org/en/v1/base/numbers/#Core.Float16)), but they are implemented in software and use [Float32](https://docs.julialang.org/en/v1/base/numbers/#Core.Float32) for calculations.
Even in CUDA half-precision floating-point support doesn't work in the same way as single- or double-precision floating point.

It is also not trivial to use half-precision floating-point in scientific algorithms due to their limited range and precision. I've added support for half-precision floating-point in our VLBI software correlator for 2-bit input samples delaying conversion to 32-bit floating-point until the accumulation stage of the correlator. This required additional normalization steps and resulted in a small coherence loss. I suspect the same implementation would not be acceptable for 8-bit samples without conversion to 32-bit floating-point at an earlier stage.

This is an area where developments go fast, and it would not surprise me if it is worth revisiting this issue in a few years as support for half-precision floating-point matures in the wider software ecosystem. There is not fundamental reason why the schema couldn't be extended to support (complex) half-precision floating point when that happens.

**Comment by giordano** [comment]
> I realize that documentation is not necessarily up to date, but even the current Julia documentation says:

Yes, that's outdated, on target with native fp16 supports native instructions are used (through the use of LLVM's [`half`](https://llvm.org/docs/LangRef.html#floating-point-types) type)

**Comment by giordano** [comment]
Thanks for pointing that out, I updated the docs in https://github.com/JuliaLang/julia/pull/57725.
